File: Search Algorithms/Java/InterpolationSearch.java

import java.util.Scanner;

public class Main {

    public static void main(String[] args) {

        int length, ele;
        Scanner scanner = new Scanner(System.in);

        System.out.print("Enter number of elements: ");
        length = scanner.nextInt();

        int arr[]=new int[length];
        System.out.println("Enter elements: ");

        for(int i=0;i<length;i++){
            arr[i]=scanner.nextInt();
        }

        System.out.println("Enter element to find: ");
        ele = scanner.nextInt();

        int index = interpolationSearch(arr, ele);

        if (index != -1)
            System.out.println("Element found at index " + index);
        else
            System.out.println("Element not found.");
    }

    public static int interpolationSearch(int[] nums, int val) {

        int lo = 0, mid = 0, hi = nums.length - 1;

        while (nums[lo] <= val && nums[hi] >= val) {

            mid = lo + ((val - nums[lo]) * (hi - lo)) / (nums[hi] - nums[lo]);
            if (nums[mid] < val) {
                lo = mid + 1;

            } else if (nums[mid] > val) {
                hi = mid - 1;

            } else return mid;
        }

        if (nums[lo] == val) return lo;

        return -1;
    }

}

File: Search Algorithms/Kotlin/InterpolationSearch.kt
fun interpolationSearch(arr: IntArray, item: Int): Int {
    var hi = arr.size - 1
    var lo = 0
    while (item >= arr[lo] && item <= arr[hi] && lo <= hi) {
        val probe =
            lo + (hi - lo) * (item - arr[lo]) / (arr[hi] - arr[lo])
        if (hi == lo) {
            return if (arr[lo] == item) {
                lo
            } else {
                -1
            }
        }
        if (arr[probe] == item) {
            return probe
        }
        if (arr[probe] < item) {
            lo = probe + 1
        } else {
            hi = probe - 1
        }
    }
    return -1
}


fun main(args: Array<String>) {
    var arr = intArrayOf(
        2, 3, 5, 6, 8, 11, 15, 19, 21, 24,
        28, 31, 44, 55, 76
    )
    val x = 31
    val xIndex = interpolationSearch(arr, x)

    if (xIndex != -1) println("Item $x found at index $xIndex") else println("Element $x not found.")

    val y = 1
    val yIndex = interpolationSearch(arr, y)

    if (yIndex != -1) println("Item $x found at index $yIndex") else println("Element $x not found.")
}
